articles/cross_compilation_build_tools_design.md

```diff
@@ -0,0 +1,161 @@
+
+Support for other platforms can be added by creating corresponding ROS 2 Docker images, and variants of the toolchain file if required.
+We could include support for ARM-HF and ARM64 in the first release of the plugin.
+
+### Frequently asked questions
+
+- _Are we building the sysroot for the workspace, and running `rosdep install` on each invocation of `colcon cc-build`?_ No, the Docker file system export is only performed once, if the sysroot is not already available. The argument `--force-sysroot-build` can be used to force building the sysroot if a sysroot file is already available, for example
+when new dependencies are added to some package in the workspace.
+- _How can we run the resulting binaries?_ Packaging the resulting binaries is outside of the scope of this document. The cross-compiled binaries will be
+available in the workspace is a directory corresponding to the architecture-OS pair, e.g. `armhf-ubuntu_bionic`. We make the assumption that the user will copy the
+whole workspace to the remote device (including src/) and run `rosdep` on the target platform, to install the dependencies before running the binaries.
+
+### Maintenance and testing
+
+Regarding maintenance, in order to support cross-compilation for a new platform we would need to:
+
+- Prepare and publish a base ROS 2 Docker image for the new platform, both fetching the dependencies using the ROS 2 source using `rosdep install`, and from the pre-built binary if available for that platform.
+- Adapt the toolchain file for the new platform, if required.
+- Extend the `cc-build` plugin to include dependencies for the compiler for the new platform, and any modifications needed to use a suitable toolchain
+file for the platform.
+
+We could then test cross-compile support for the new platform by building the base  ROS 2 Docker image, and a custom sysroot image for some reference package
+(e.g. [ApexAI/performance_test](https://github.com/ApexAI/performance_test/)), and then launching a Docker container for the that sysroot image, running
+`colcon test` using the cross-compiled binaries. The container would be launched with a [bind mount](https://docs.docker.com/storage/bind-mounts/) of the
+workspace so they have access to the compiled binaries.
+
+In order to simplify the development of base ROS 2 Docker images for new architecture-OS combinations, on a second iteration of this plugin we might add a
+new verb `colcon cc-build-sysroot-base-image` that will 1) copy the required qemu files and ROS 2 source into a temporary file; 2) build a Docker image from a
+specified Dockerfile, with access to the those resources; 3) optionally publish the image into a Docker registry, using a suitable naming convention like
+`${docker_repo_uri}:${os}_${arch}_${prebuilt?}_${rosdistro}`. Dockerfiles compatible with this command, for each architecture-OS pair would be keep on the
+`ros2/cross_compile` repository, so they can be used in CI/CI pipeline if needed.
```

Comment:
CI/CI pipeline -> CI/C**D** pipeline ?

Reply:
I'll fix that, thanks

Reply:
I don't know about you but I feel the need to do CI on my CI all the time. Who watches the watchers and all that.